src/ThreeWayMerge.php

```diff
@@ -0,0 +1,206 @@
+<?php
+
+namespace Relaxed\Merge\ThreeWayMerge;
+
+use Exception;
+
+class ThreeWayMerge
+{
+    public function performMerge(array $ancestor, array $local, array $remote)
+    {
+        // Returning a new Array for now. Can return the modified ancestor as well.
+        $merged = [];
+        foreach ($ancestor as $key => $value) {
+            if (is_array($value)) {
+                $merged[$key] = $this->performMerge(
+                    $value,
+                    $local[$key],
+                    $remote[$key]
+                );
+            } else {
+                $merged[$key] = $this->merge(
+                    $ancestor[$key],
+                    $local[$key],
+                    $remote[$key],
+                    $key
+                );
+                if ($merged[$key] == null) {
+                    unset($merged[$key]);
+                }
+            }
+        }
+        return $merged;
+    }
+
+    protected function merge($x, $y, $z, $key)
+    {
+        $ancestor = (strpos($x, "\n") !== false ? explode("\n", $x) : array($x));
+        $local = (strpos($y, "\n") !== false ? explode("\n", $y) : array($y));
+        $remote = (strpos($z, "\n") !== false ? explode("\n", $z) : array($z));
+        $count_ancestor = count($ancestor);
+        $count_remote = count($remote);
+        $count_local = count($local);
+        if ($count_ancestor != $count_local
+            && $count_local!= $count_remote
+            && $count_ancestor != $count_remote
+        ) {
+            $merged = $this->linesModified(
+                $ancestor,
+                $local,
+                $remote,
+                $count_ancestor,
+                $count_local,
+                $count_remote
+            );
+        } else {
+            $count = $count_ancestor == $count_local ? $count_remote : $count_local;
+            $merged = $count > $count_ancestor ?
+                $this->linesAdded(
+                    $ancestor,
+                    $local,
+                    $remote,
+                    $count,
+                    $count_remote,
+                    $count_ancestor,
+                    $count_local
+                ) :
+                $this->linesRemovedOrModified($ancestor, $local, $remote, $count);
```

Comment:
To improve readability I think sections like this one should be broken out into full `if` `else` statements, instead of the shorthand versions of it.